Bugfix/split diagnostics files

pchila opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

Move the full components definition and state from `state.yaml` into a separate diagnostic file `components_actual.yaml` 
In `state.yaml` we can still find the state of the components and the units but now it should be easier to read.
<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

## Why is it important?

Helps with incidents/bugs investigation since it's now easier to grasp the state of the agent and the components while maintaining full details of the running component in a separate file

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->
Launch an agent and execute the `elastic-agent diagnostics` to generate the archive with the new files contents
## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes #2139 
- Supersedes #2350  (those commits are included here, we can decide if we want to review the 2 PRs separately or just merge this one)
## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-2380/detail/PR-2380/2//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-2380/detail/PR-2380/2//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-2380/detail/PR-2380/2//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-2380/detail/PR-2380/2//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-agent_2380.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-03-21T17:01:20.186Z&rangeTo=2023-03-21T17:21:20.186Z&transactionName=BUILD+elastic-agent%2Felastic-agent-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=2ac6a7c26180e64493da5046dd99a43c&transactionId=18641feb5452fd15)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-03-21T17:11:20.186+0000


* Duration: 17 min 30 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 5023  |
| Skipped      | 15 |
| Total        | 5038   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages.


- `run integration tests` : Run the Elastic Agent Integration tests.


- `run end-to-end tests` : Generate the packages and run the E2E Tests.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 98.387% (`61/62`) | :+1: 
Files | 69.34% (`147/212`) | :+1: 
Classes | 68.473% (`278/406`) | :+1: 
Methods | 53.668% (`856/1595`) | :+1: 0.029
Lines | 39.135% (`9556/24418`) | :+1: 0.204
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

pchila commented:
@cmacknz  It would be great if you could have a look at least at the expected .yaml files to check if this change is what you expect out of #2139 
(keep in mind that the unit and unit states are missing from those because we are not able to create/sanitize the hook output data properly yet but they are dumped by the diagnostic command)

cmacknz commented:
Took a look at the output. Initial feedback:

* state.yaml looks great
* components_actual.yaml has some extra indentation that can be removed.

Here are the first few lines of components_actual.yaml:
```yaml
            components:
            - component:
                id: system/metrics-default
                input_spec:
                  binary_name: metricbeat
                  binary_path: ../../../../../specs/metricbeat
                  input_type: system/metrics
```

vs components_expected.yaml:

```yaml
components:
- id: log-default
  input_spec:
    binary_name: filebeat
    binary_path: ../../../../../specs/filebeat
    input_type: log
```
* components_actual.yaml still contains a `state:` key and it looks like it contains mock information. We should remove the `state:` key entirely if we can:
```
              state:
                message: ""
                state: 2
                units: {}
                version_info:
                  name: Mock system/metrics
                  version: 1.2.3
```
* components_actual.yaml seems to be missing the actual unit configuration. It lists the units but omits their configuration:
```yaml
                units:
                - id: system/metrics-default-system/metrics-system-d38b6b0b-cc52-4a61-b897-1b236982ac42
                  log_level: 2
                  type: 0
                - id: system/metrics-default
                  log_level: 2
                  type: 1
```

For reference this is included in components_expected.yaml, and I think we should keep it so we have the actual applied configuration in the diagnostics:

```yaml
  - config:
      datastream: null
      id: ""
      meta: null
      name: ""
      revision: 0
      source:
        fields:
          hosts:
            kind:
              listvalue:
                values:
                - kind:
                    stringvalue: https://your.elasticsearch.host:443
          type:
            kind:
              stringvalue: elasticsearch
      streams: []
      type: elasticsearch
    id: log-default
    log_level: 2
    type: 
```
* We aren't using consistent separators for multi-word file names, I don't have a strong preference beyond that it be consistent. I suggest using `-` instead of `_` since that is what was used in the filenames that existed before this change:
```sh
❯ ls diag
allocs.txt               components_expected.yaml logs                     threadcreate.txt
block.txt                computed-config.yaml     mutex.txt                variables.yaml
components               goroutine.txt            pre-config.yaml          version.txt
components_actual.yaml   heap.txt                 state.yaml
```

cmacknz commented:
Also, as follow up work we will need to update the list of files in the diagnostics documentation: https://www.elastic.co/guide/en/fleet/current/elastic-agent-cmd-options.html#elastic-agent-diagnostics-command

The source for this page is in https://github.com/elastic/observability-docs/blob/main/docs/en/ingest-management/commands.asciidoc#elastic-agent-diagnostics-command

I believe there is also a support knowledge base article we need to update but I won't link it here because it isn't publicly accessible.

pchila commented:
@cmacknz I implemented the changes about the components file names and removed the `state` information from the `components-actual.yaml` file.

The problem with units and their configuration comes from the datamodel we use and mocking:
- The units configuration in the `components-actual.yaml` is a protobuf object returned by the Runtime (which we mock away) and that is proving challenging to create without lengthy and clumsy sections of code. 
I tried parsing it from yaml contained in `components_units.yaml`:
    - using [`component.ExpectedConfig()`](https://github.com/elastic/elastic-agent/blob/c8d846753d6f76eec6fe962e9647357376e2e2b3/pkg/component/config.go#L36) function but it seems to have issues when setting the `Source` field.
    - using [`RuntimeSpecs.PolicyToComponents()`](https://github.com/elastic/elastic-agent/blob/c8d846753d6f76eec6fe962e9647357376e2e2b3/pkg/component/component.go#L146) is also problematic since it will instantiate its own SysInfo vars and because of that will produce different results on different systems regardless of what the unit test sets (it could be refactored to accepts vars from outside though)
    - using JSON representing the protobuf `UnitExpectedConfig` and then unmarshaling using [`protojson.Unmarshal`](https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson#Unmarshal) has some issues with size (we have sometimes 15KB+ minified JSON strings to store somewhere + the unmarshaler throws an EOF error if we try to unmarshal the unit config as returned by a `diagnostics` command)
    
- Unit states are easier but we use a composite key (unit type, unit ID) which makes the generated YAML unparseable by the `[yaml.v2](https://pkg.go.dev/gopkg.in/yaml.v2)` package (did brefly experiment with the `[yaml.v3](https://pkg.go.dev/gopkg.in/yaml.v3)` and it looked promising, however there are a lot of other changes that it would bring in so I didn't feel comfortable in using in the coordinator test code). This makes impossible to sanitize hook output for the test purposes.

The issues we have with unit configuration and state make mocking realistic unit data very difficult unless we resort to pulling in more and more components (like a real Runtime for example) which would sort of trivialize and remove effectiveness from the tests (we are basically comparing the output of coordinator with an output calculated using a lot of the same functionality as the system under test so they will be equally correct/wrong).

The alternative is to use simplified unit configs and state in the test but I feel that that would also detract from the tests usefulness.

@cmacknz please feel free to advise if you see a viable solution to the unit mocking problems listed above.

cmacknz commented:
> using [component.ExpectedConfig()](https://github.com/elastic/elastic-agent/blob/c8d846753d6f76eec6fe962e9647357376e2e2b3/pkg/component/config.go#L36) function but it seems to have issues when setting the Source field.

What are the issues? Is this actually a bug ;)

> using [RuntimeSpecs.PolicyToComponents()](https://github.com/elastic/elastic-agent/blob/c8d846753d6f76eec6fe962e9647357376e2e2b3/pkg/component/component.go#L146) is also problematic since it will instantiate its own SysInfo vars and

Can we just ignore the keys we variables.yaml we know will vary per platform?

> Unit states are easier but we use a composite key (unit type, unit ID) which makes the generated YAML unparseable by the [yaml.v2](https://pkg.go.dev/gopkg.in/yaml.v2) package ( I did brefly experiment with the [yaml.v3](https://pkg.go.dev/gopkg.in/yaml.v3) and it looked promising, however there are a lot of other changes that it would bring in so I didn't feel comfortable in using in the coordinator test code)

We probably need to move to yaml.v3 eventually, but that is out of scope for this PR. Can we change the data type we write to disk to not use a composite key? These YAML files aren't an API contract so we can change them.

cmacknz commented:
I installed the latest changes locally and the changes in the diagnostics LGTM.

pchila commented:
@cmacknz Thanks for the feedback! :) 

> > using [component.ExpectedConfig()](https://github.com/elastic/elastic-agent/blob/c8d846753d6f76eec6fe962e9647357376e2e2b3/pkg/component/config.go#L36) function but it seems to have issues when setting the Source field.
> 
> What are the issues? Is this actually a bug ;)

It could be or just be improper use on my side, maybe we can improve on it in a followup PR ?

> > using [RuntimeSpecs.PolicyToComponents()](https://github.com/elastic/elastic-agent/blob/c8d846753d6f76eec6fe962e9647357376e2e2b3/pkg/component/component.go#L146) is also problematic since it will instantiate its own SysInfo vars and
> 
> Can we just ignore the keys we variables.yaml we know will vary per platform?
> 
We could but the rendering of the components/units will be different when running on a linux or windows or mac machine for a basic policy I took from a running fleet. I also tried getting that out of the equation (with limited success) so that, whatever the OS, the unit test always have the same output

> > Unit states are easier but we use a composite key (unit type, unit ID) which makes the generated YAML unparseable by the [yaml.v2](https://pkg.go.dev/gopkg.in/yaml.v2) package ( I did brefly experiment with the [yaml.v3](https://pkg.go.dev/gopkg.in/yaml.v3) and it looked promising, however there are a lot of other changes that it would bring in so I didn't feel comfortable in using in the coordinator test code)
> 
> We probably need to move to yaml.v3 eventually, but that is out of scope for this PR. Can we change the data type we write to disk to not use a composite key? These YAML files aren't an API contract so we can change them.

If the unitID is enough to uniquely identify the unit  this would be possible, however I have to check if it doesn't change too much the rest of the code: we do not have a separate datamodel for diagnostics so this change would impact the internals of agent as well :disappointed:

blakerouse reviewed (approved):
Change looks good to me.
